src/architecture.jl

```diff
@@ -0,0 +1,13 @@
+"""
+Abstract supertype for architectures supported by DFTK.
+"""
+abstract type AbstractArchitecture end
+
+struct CPU <: AbstractArchitecture end
+
+get_array_type(::CPU) = Array
+
+"""
+Generic, hardware independent architecture for DFTK.
+"""
+abstract type GPU <: AbstractArchitecture end
```

Comment:
@vchuravy Are you aware of a more wide-spread solution for this type of "device / computing architecture selection". It seems currently everyone is rolling their own ... including naming conflicts for `GPU`, `CPU` etc. I would like to avoid this, but this leads to a pretty powerful and extensible interface to specify hardware setup.

Reply:
Yeah, I have some of that in KA, https://github.com/JuliaGPU/KernelAbstractions.jl/pull/320 and there is https://docs.juliahub.com/GPUArraysCore/qiYUe/0.1.2/autodocs/#GPUArraysCore.backend-Tuple{Type}

@oschulz was working on something https://github.com/JuliaGPU/KernelAbstractions.jl/pull/297

cc: @maleadt @jpsamaroo We do need some sort of "Query compute backend for <:AbstractArray" and "allocate array for backend (similar is not always enough)"